youths/views.py

```diff
@@ -1,16 +1,48 @@
 from django.conf import settings
-from django.http import HttpResponse, JsonResponse
+from django.http import HttpResponse
+from django.shortcuts import get_object_or_404
+from rest_framework import serializers, status
+from rest_framework.response import Response
+from rest_framework.views import APIView
 
 from youths.models import YouthProfile
 
 
-def profiles(request, *args, **kwargs):
-    if settings.GDPR_API_ENABLED:
-        # TODO: Add authentication and security
-        if request.method == "GET":
-            youth_profile = YouthProfile.objects.get(profile__pk=kwargs["id"])
-            return JsonResponse(youth_profile.serialize(), safe=False)
-        else:
-            return HttpResponse(status=405)
-    else:
-        return HttpResponse(status=404)
+class DryRunSerializer(serializers.Serializer):
+    dry_run = serializers.BooleanField(required=False, default=False)
+
+
+class YouthProfileGDPRAPIView(APIView):
+    # TODO: Add authentication and security
+
+    def dispatch(self, request, *args, **kwargs):
+        if not settings.GDPR_API_ENABLED:
+            return HttpResponse(status=404)
+        return super().dispatch(request, *args, **kwargs)
+
+    def get_object(self) -> YouthProfile:
+        return get_object_or_404(YouthProfile, profile__pk=self.kwargs["pk"])
+
+    def is_dry_run(self):
+        """Check if parameters provided to the view indicate it's being used for dry_run."""
+        data = DryRunSerializer(data=self.request.data)
+        query = DryRunSerializer(data=self.request.query_params)
+        data.is_valid()
+        query.is_valid()
+
+        return data.validated_data["dry_run"] or query.validated_data["dry_run"]
+
+    def get(self, request, *args, **kwargs):
+        """Retrieve all youth profile data related to the given id."""
+        return Response(self.get_object().serialize(), status=status.HTTP_200_OK)
+
+    def delete(self, request, *args, **kwargs):
+        """Delete all data related to the given youth profile.
+
+        Deletes all data related to the given profile id, or just checks if the data can be deleted,
+        depending on the `dry_run` parameter.
+        """
+        if not self.is_dry_run():
+            self.get_object().delete()
```

Comment:
If we had some reason why particular youth profile could not be removed how would that be implemented? The initial plan was to use the `dry_run` parameter in such a way that we simulate the deletion without actually committing it. I guess the easiest (and most DRY) way to accomplish this would be doing deletion in transaction and then in the end of the dry run just by calling `transaction.rollback()` to rollback the deletion.

For youth membership I'm not sure will there ever be a situation where the `YouthMembership` model cannot be deleted but at least for Venepaikka there is already use case for that and for that reason it would be nice to already have a placeholder for that logic because I think the idea is to provide with a copy-paste example or probably even a lib solution for other projects to use.

Reply:
I added some transaction related logic into the view. Though maybe for a library like functionality there probably should also be some business logic call to cover cases where `DatabaseError` isn't enough to determine if something can be deleted (e.g. there could a relevant piece of data that is not protected by ForeignKey's `models.PROTECT`).

Reply:
Agreed, but it doesn't have to be a complete library yet, just keeping our heading towards that. But in general it would be nice to use db constraints as much as possible to avoid writing too many custom checks and codes. Especially having so many different API admin, GraphQL, REST, etc. I'd keep this kind of stuff as low level as possible of course keeping options open for custom checks.